libs/qCC_io/ShpFilter.cpp

```diff
@@ -721,68 +701,74 @@ static std::vector<CCVector3> readPoints(QDataStream& shpStream, int32_t numPoin
 		points[i].y = static_cast<PointCoordinateType>(y + Pshift.y);
 		points[i].z = 0;
 	}
-	return points;
+
+	return CC_FERR_NO_ERROR;
 }
 
-static std::vector<ScalarType> readMeasures(QDataStream& shpStream, int32_t numPoints)
+static CC_FILE_ERROR ReadMeasures(QDataStream& shpStream, int32_t numPoints, std::vector<ScalarType>& scalarValues)
 {
-	std::vector<ScalarType> scalarValues;
 	//M boundaries
 	double mMin;
 	double mMax;
 	shpStream >> mMin >> mMax;
 
-	if (!isESRINoData(mMin) && !isESRINoData(mMax))
+	if (IsESRINoData(mMin) || IsESRINoData(mMax))
 	{
-		try
-		{
-			scalarValues.resize(numPoints);
-		}
-		catch (const std::bad_alloc&)
-		{
-			return scalarValues;
-		}
+		ccLog::Warning("[SHP] Invalid measurements range");
+		return CC_FERR_MALFORMED_FILE;
 	}
 
-	//M values (an array of length NumPoints)
-	if (!scalarValues.empty())
+	try
 	{
-		for (int32_t i = 0; i < numPoints; ++i)
-		{
-			double m;
-			shpStream >> m;
-			scalarValues[i] = (isESRINoData(m)  ? NAN_VALUE : static_cast<ScalarType>(m));
-		}
+		scalarValues.resize(numPoints);
 	}
-	else
+	catch (const std::bad_alloc&)
+	{
+		return CC_FERR_NOT_ENOUGH_MEMORY;
+	}
+
+	//M values (an array of length NumPoints)
+	for (int32_t i = 0; i < numPoints; ++i)
 	{
-		shpStream.skipRawData(numPoints * 8);
+		double m;
+		shpStream >> m;
+		scalarValues[i] = (IsESRINoData(m) ? NAN_VALUE : static_cast<ScalarType>(m));
 	}
-	return scalarValues;
+
+	return CC_FERR_NO_ERROR;
 }
 
 //! Builds the ccPointCloud of vertices
-//! Returns nullptr if out of memory
-ccPointCloud *buildVertices(const std::vector<CCVector3> &points, int32_t firstIndex, int32_t lastIndex)
+/** \return nullptr if out of memory
+**/
+static ccPointCloud* BuildVertices(const std::vector<CCVector3> &points, int32_t firstIndex, int32_t lastIndex)
 {
 	int32_t vertCount = lastIndex - firstIndex + 1;
+	if (vertCount <= 0)
+	{
+		assert(false);
+		return nullptr;
+	}
 
-	auto *vertices = new ccPointCloud("vertices");
+	ccPointCloud* vertices = new ccPointCloud("vertices");
```

Comment:
I thought when we discussed `auto` way back in the C++11 change, this was acceptable?

Generally this is an acceptable "new C++ style" that [clang static analysis](https://clang.llvm.org/extra/clang-tidy/checks/modernize-use-auto.html) can warn about.

It avoids repeating the type that's already there and makes for easier-to-read code.

```cpp
auto foo = 5; //bad;
auto bar = new blat( "ding" ); //good
```

Reply:
I personally think it's less explicit (you have to look at the code on the right to understand the type. And it's particularly 'lazy' when it's not for very long templated types.

Reply:
You're never going to _only_ read `auto *vertices` or `ccPointCloud* vertices` and then stop. You're going to read the whole line anyway, so it's just extra noise.

Reply:
Not always. Depends on the circumstances. Really, for me auto is a shortcut that should only be used when necessary.

Reply:
I mean, if you only want to know the type, you'll get the information faster.